### PR TITLE
kexec: Decompress gzip compressed kernels

### PR DESCRIPTION
**Pull request by rjoleary**
For arm64, make simply builds a Image and Image.gz. Not self
decompression. The bootloader (like u-boot and others) are expected to
decompress.

Signed-off-by: Ryan O'Leary <ryanoleary@google.com>

### CODE REVIEW

File: pkg/boot/kexec/kexec_linux_implemented.go [hunk 1]
```diff
@@ -26,8 +34,38 @@ func FileLoad(kernel, ramfs *os.File, cmdline string) error {
...
 	}
 
+	if isGzip(kernel) {
+		g, err := gzip.NewReader(kernel)
+		if err != nil {
+			return fmt.Errorf("could read kernel gzip header: %v", err)
+		}
+		decompressed, err := ioutil.TempFile("", "kernel")
+		if err != nil {
+			return fmt.Errorf("could not create temp file: %v", err)
```
Comment (rminnich) [inline pkg/boot/kexec/kexec_linux_implemented.go:44]: could not create temp file for decompression
?
people might get confused as to what the context is

File: pkg/boot/kexec/kexec_linux_implemented.go [hunk 2]
```diff
@@ -26,8 +34,38 @@ func FileLoad(kernel, ramfs *os.File, cmdline string) error {
...
+		defer os.Remove(decompressed.Name())
+		if _, err := io.Copy(decompressed, g); err != nil {
+			return fmt.Errorf("could not decompress kernel: %v", err)
+		}
+		if err := g.Close(); err != nil {
+			return fmt.Errorf("could not close gunzip (bad checksum?): %v", err)
+		}
+		if err := decompressed.Sync(); err != nil {
+			return fmt.Errorf("could not sync decompressed kernel: %v", err)
+		}
```
Comment (rminnich) [inline pkg/boot/kexec/kexec_linux_implemented.go:55]: printing the name "might" be helpful as knowing the path might indicate the problem, e.g. /tmp was out of room so the sync failed.

File: pkg/boot/kexec/kexec_linux_implemented.go [hunk 3]
```diff
@@ -26,8 +34,38 @@ func FileLoad(kernel, ramfs *os.File, cmdline string) error {
...
+		kernel = decompressed
+	}
+
 	if err := unix.KexecFileLoad(int(kernel.Fd()), ramfsfd, cmdline, flags); err != nil {
 		return fmt.Errorf("sys_kexec(%d, %d, %s, %x) = %v", kernel.Fd(), ramfsfd, cmdline, flags, err)
 	}
 	return nil
 }
+
+// isGzip returns true if the file is gzip file format.
```
Comment (rminnich) [inline pkg/boot/kexec/kexec_linux_implemented.go:65]: amazed that this is not in the package ...

**Review by rminnich — APPROVED** [review]

File: pkg/boot/kexec/kexec_linux_implemented.go (hunk 1 above)
Comment (hugelgupf) [inline pkg/boot/kexec/kexec_linux_implemented.go:44]: If any of these fail, I'd try kexec_file_load anyway, especially on x86.

File: pkg/boot/kexec/kexec_linux_implemented.go (hunk 3 above)
Comment (rjoleary) [reply pkg/boot/kexec/kexec_linux_implemented.go:65]: You should see how complex the C version is: https://github.com/horms/kexec-tools/blob/master/kexec/arch/arm64/kexec-zImage-arm64.c

**Comment by rminnich** [comment]
so ... you gonna go with this or ... ?

**Comment by rjoleary** [comment]
It looks like this functionality was already merged in https://github.com/u-root/u-root/pull/1828

I'll have to see why it didn't work for my Linux kernel (received Invalid Argument for kexec syscall).

**Review by rminnich — APPROVED** [review]

**Comment by rminnich** [comment]
oh. I kind of did think I'd done this. Hm. Let me know what you want to do.

**Pull request closed without merging**